playbooks/common/openshift-cluster/std_include.yml

```diff
@@ -13,28 +13,22 @@
       aggregate: false
 
 - include: evaluate_groups.yml
-  tags:
-  - always
 
 - include: initialize_facts.yml
-  tags:
-  - always
 
 - include: sanity_checks.yml
-  tags:
-  - always
 
 - include: validate_hostnames.yml
-  tags:
-  - node
 
 - include: initialize_openshift_repos.yml
-  tags:
-  - always
+
+# openshift_setup_docker should be set a include task's variable.
+# We only need to run setup_docker on first cluster install.
+# We set openshift_setup_docker to True in playbooks/byo/openshift-cluster/config.yml
+- include: setup_docker.yml
```

Comment:
I like where you are going with this PR.  I like the fact that docker can be pulled out but I'm not sure `std_include.yml` is the right place for it.  `std_include.yml` should only be for initialization of the bare minimum for someone to call our roles.

@mtnbikenc, Any thoughts on where this docker include should go?

Reply:
Yeah, this is kind of a compromise.  I'm trying to do the least amount of disruption as possible here.  We need the docker role to be executed before openshift_version and health_checks.  Since openshift_version is in std_include, it has to be here unless I want to break up std_include for every entry point play.  I'm not opposed to that approach, but I don't think this is the patch set to do it in.

Effectively, we're executing docker for containerized hosts via openshift_version in std_include today.  The only change here is we're aligning all hosts to execute the docker role during this time.  Further, we're only running docker during the initial install (or rerun of byo/config.yml).

Reply:
Installing docker, for any host type, during std_include is not the right path going forward.  We should refactor openshift_version to remove the dependency on docker.  We are only using docker to check for image availability, not starting containers.  We have investigated other options for checking for image availability in registries.  With the option of using CRI-O, we should not prolong the requirement of installing docker to check registries.

Reply:
To follow on to @mtnbikenc's comments note that for 3.7's use of cri-o will still have docker running. Though work is being done to allow a [100% cri-o experience](https://trello.com/c/4A9jTY67/523-1-implement-openshiftcriosystemcontainerimageoverride#comment-59d259296a2fe65c661b8d8d)

Reply:
We should definitely refactor openshift_version.  Unfortunately, we're past feature freeze for 3.7, and we rely on it today.

This patch needs to ship to fix the linked BZ to ship our RFE for registry authentication.  If we don't implement some kind of fix to how the docker role is installed, then we have to tell users of the registry authentication to disable docker image availability checks or the new feature is going to fail for RPM based installs.

I say we ship this, refactor openshift_version role later.  That role sets facts that are used all over the place, lots of logic hinges on that role executing properly.  I don't want to block registry authentication on refactoring openshift_version.

Reply:
The bug references the failure of docker_image_availability checking authenticated registries.  The checker uses skopeo so the checker should be updated to pass the creds for authenticated registries if they are needed unless I'm missing something else on this one.
https://github.com/projectatomic/skopeo#private-registries-with-authentication
@sosiouxme @ashcrow Thoughts?

Reply:
I grep'd the code for ``.docker/config.json`` so it looks like the proper configuration doesn't get created for auto sensing configuration. I would agree with @mtnbikenc's statement and credentials should be provided.

Reply:
If we can fix this by simply passing credentials to the image availability check and defer this refactoring I think that's the best path forward given the timing. We should take refactoring this and openshift_version not to require docker on as a card.

Reply:
.docker/config.json is created by the docker role, using the 'docker login' command.  That's why I'm trying to get the docker role to fire first, that's where those credentials are created.

The password in $HOME/.docker/config.json is not the same the password provided by the user.  This password is a token that is generated using the username/password combo provided by the user.

However, looks like skopeo might work with that username/password combo for the purposes of the health checks.

I'll cut a new patch, if I can pass the creds to skopeo without a bunch of hassle, then we can table this patch for now.

Reply:
Etcd v2->v3 migration and soon embedded etcd -> external etcd migration should not rely on the docker role at all. Both plays do/will include the `std_include.yml` and I would really like to run only the minimum set of tasks I need to run the migration. I am not saying what we should do now, just saying that at the end it would be really awesome to run only what we need to.

Reply:
We have some work at https://github.com/openshift/openshift-ansible/pull/5650 to use auth on `oreg_url`.

Is there a need to provide auth for other registries, and a way to determine the credentials?

In an ideal world, the docker role would run to determine all necessary information (not necessarily configure anything) and the `docker_image_availability` check would just consume that information rather than duplicate the logic.

Reply:
At this point, I'm starting to feel getting the skopeo image check pieces working correctly is more error-prone than this work.

If docker is already configured, we don't have to do anything special in the image check.  The required credentials will already be on disk, and skopeo will use them.

Reply:
Had I taken a moment to read https://github.com/projectatomic/skopeo#private-registries-with-authentication - skopeo should just work once docker login has been performed as necessary. So yeah, that's ideal.

Reply:
Keep in mind using ``docker`` to check for images will end up not being sufficient, but that will be in a future release. For the current release either works.